File: pset1/credit/credit.c
#include <cs50.h>
#include <math.h>
#include <stdio.h>

int Luhn_algorithm_valid(long number);
int find_digit(long number, int i);
int get_size(long number);
int american_express_valid(long number);
int mastercard_valid(long number);
int visa_valid(long number);

int main(void)
{
    long number;
    do
    {
        number = get_long("Enter your credit card number: ");
    }
    while (number < 0);

    if (Luhn_algorithm_valid(number))
    {
        if (american_express_valid(number))
        {
            printf("AMEX\n");
        }
        else if (mastercard_valid(number))
        {
            printf("MASTERCARD\n");
        }
        else if (visa_valid(number))
        {
            printf("VISA\n");
        }
        else
        {
            printf("INVALID\n");
        }
    }
    else
    {
        printf("INVALID\n");
    }
}

int american_express_valid(long number)
{
    int size = get_size(number);
    int first_digit = find_digit(number, size);
    int second_digit = find_digit(number, size - 1);
    return (size == 15 && first_digit == 3 && (second_digit == 4 || second_digit == 7));
}

int mastercard_valid(long number)
{
    int size = get_size(number);
    int first_digit = find_digit(number, size);
    int second_digit = find_digit(number, size - 1);
    return (size == 16 && first_digit == 5 &&
            (second_digit == 1 || second_digit == 2 || second_digit == 3 || second_digit == 4 || second_digit == 5));
}

int visa_valid(long number)
{
    int size = get_size(number);
    int first_digit = find_digit(number, size);
    int second_digit = find_digit(number, size - 1);
    return ((size == 13 || size == 16) && first_digit == 4);
}

int Luhn_algorithm_valid(long number)
{
    int sum = 0;
    for (int i = 1; i <= get_size(number); i++)
    {
        if (i % 2 == 0)
        {
            int digit = find_digit(number, i) * 2;
            for (int j = 1; j <= get_size(digit); j++)
            {
                sum += find_digit(digit, j);
            }
        }
        else
        {
            int digit = find_digit(number, i);
            sum += digit;
        }
    }
    return sum % 10 == 0;
}

int find_digit(long number, int i)
{
    if (i > get_size(number))
    {
        return 0;
    }
    if (i == 1)
    {
        return number % 10;
    }
    return ((number % (long long) pow(10, i))) / (long long) pow(10, i - 1);
}

int get_size(long number)
{
    if (number == 0)
    {
        return 1;
    }
    int size = 0;
    while (number != 0)
    {
        number /= 10;
        size++;
    }
    return size;
}

File: pset6/sentimental/credit/credit.py
from cs50 import get_int


def main():
    while True:
        number = get_int("Number: ")
        if number > 0:
            break

    if luhnAlgorithm(number):
        print(f"{issuer(number)}")
    else:
        print("INVALID")


def luhnAlgorithm(number):
    sum = 0
    cursor = 1
    snumber = str(number)[::-1]

    for i in snumber:
        if cursor % 2 == 0:
            digit = int(i) * 2
            sdigit = str(digit)
            for c in sdigit:
                sum += int(c)
        else:
            digit = int(i)
            sum += digit
        cursor += 1

    return sum % 10 == 0


def issuer(number):
    string = str(number)
    size = len(string)
    first = int(string[0])
    second = int(string[1])

    if size == 15 and first == 3 and (second == 4 or second == 7):
        return "AMEX"
    elif size == 16 and first == 5 and (second in [1, 2, 3, 4, 5]):
        return "MASTERCARD"
    elif (size == 13 or size == 16) and first == 4:
        return "VISA"

    return "INVALID"


main()
